File: 프로그래머스/2/468379. 선인장 숨기기/선인장 숨기기.py
from collections import deque

def solution(m, n, h, w, drops):
    INF = len(drops) + 1

    rain = [[INF] * n for _ in range(m)]

    for t, (r, c) in enumerate(drops, start=1):
        rain[r][c] = t

    row_min = [[0] * (n - w + 1) for _ in range(m)]

    for r in range(m):
        dq = deque()

        for c in range(n):
            while dq and rain[r][dq[-1]] > rain[r][c]:
                dq.pop()

            dq.append(c)

            if dq[0] <= c - w:
                dq.popleft()

            if c >= w - 1:
                row_min[r][c - w + 1] = rain[r][dq[0]]

    area_min = [[0] * (n - w + 1) for _ in range(m - h + 1)]

    for c in range(n - w + 1):
        dq = deque()

        for r in range(m):
            while dq and row_min[dq[-1]][c] > row_min[r][c]:
                dq.pop()

            dq.append(r)

            if dq[0] <= r - h:
                dq.popleft()

            if r >= h - 1:
                top = r - h + 1
                area_min[top][c] = row_min[dq[0]][c]

    best = -1
    answer = [0, 0]

    for r in range(m - h + 1):
        for c in range(n - w + 1):
            if area_min[r][c] > best:
                best = area_min[r][c]
                answer = [r, c]

    return answer

File: 프로그래머스/2/468379. 선인장 숨기기/선인장 숨기기.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> solution(int m, int n, int h, int w, vector<vector<int>> drops) {
    int INF = (int)drops.size() + 1;

    vector<vector<int>> rain(m, vector<int>(n, INF));

    for (int t = 0; t < drops.size(); t++) {
        int r = drops[t][0];
        int c = drops[t][1];
        rain[r][c] = t + 1;
    }

    vector<vector<int>> rowMin(m, vector<int>(n - w + 1));

    for (int r = 0; r < m; r++) {
        deque<int> dq;

        for (int c = 0; c < n; c++) {
            while (!dq.empty() && rain[r][dq.back()] > rain[r][c]) {
                dq.pop_back();
            }

            dq.push_back(c);

            if (dq.front() <= c - w) {
                dq.pop_front();
            }

            if (c >= w - 1) {
                rowMin[r][c - w + 1] = rain[r][dq.front()];
            }
        }
    }

    vector<vector<int>> areaMin(m - h + 1, vector<int>(n - w + 1));

    for (int c = 0; c < n - w + 1; c++) {
        deque<int> dq;

        for (int r = 0; r < m; r++) {
            while (!dq.empty() && rowMin[dq.back()][c] > rowMin[r][c]) {
                dq.pop_back();
            }

            dq.push_back(r);

            if (dq.front() <= r - h) {
                dq.pop_front();
            }

            if (r >= h - 1) {
                int top = r - h + 1;
                areaMin[top][c] = rowMin[dq.front()][c];
            }
        }
    }

    int best = -1;
    vector<int> answer = {0, 0};

    for (int r = 0; r < m - h + 1; r++) {
        for (int c = 0; c < n - w + 1; c++) {
            if (areaMin[r][c] > best) {
                best = areaMin[r][c];
                answer = {r, c};
            }
        }
    }

    return answer;
}